File: native/android/app/src/main/java/com/nemoyu/wheretostudy/nativeapp/ShuttleBusClient.kt
package com.nemoyu.wheretostudy.nativeapp

import android.os.Handler
import android.os.Looper
import java.net.URI
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.atomic.AtomicBoolean
import org.json.JSONObject

data class ShuttleBusService(
    val vehicle: String,
    val count: Int,
)

data class ShuttleBusRow(
    val departureTime: String,
    val services: Map<String, ShuttleBusService?>,
)

data class ShuttleBusPeriod(
    val label: String,
    val startDate: String?,
    val endDate: String?,
)

data class ShuttleBusSchedule(
    val period: ShuttleBusPeriod,
    val from: String?,
    val to: String?,
    val parseStatus: String,
    val rows: List<ShuttleBusRow>,
)

data class ShuttleBusNotice(
    val id: String,
    val title: String,
    val publishedAt: String,
    val sourceURL: String?,
    val kind: String,
    val parseStatus: String,
    val stops: List<Pair<String, String>>,
    val notes: List<String>,
    val schedules: List<ShuttleBusSchedule>,
)

data class ShuttleBusSnapshot(
    val generatedAt: String,
    val status: String,
    val sourceName: String,
    val sourcePage: String,
    val notices: List<ShuttleBusNotice>,
)

data class TodayShuttleDeparture(
    val time: String,
    val vehicle: String,
    val count: Int,
)

data class TodayShuttleRoute(
    val from: String,
    val to: String,
    val periodLabel: String,
    val departures: List<TodayShuttleDeparture>,
)

data class TodayShuttlePresentation(
    val status: String,
    val nextDeparture: String?,
    val noticeTitle: String?,
    val noticeURL: String?,
    val stops: List<Pair<String, String>>,
    val notes: List<String>,
    val routes: List<TodayShuttleRoute>,
    val isStale: Boolean,
)

internal object ShuttleBusResponseParser {
    private val weekdays = setOf(
        "monday", "tuesday", "wednesday", "thursday",
        "friday", "saturday", "sunday",
    )
    private val timePattern = Regex("^(?:[01]\\d|2[0-3]):[0-5]\\d$")

    fun parse(payload: String): ShuttleBusSnapshot {
        val root = try {
            JSONObject(payload)
        } catch (error: Exception) {
            throw DailyInfoClientException("班车数据格式不正确。", error)
        }
        if (root.optString("schema_version") != "1.0") {
            throw DailyInfoClientException("班车数据版本不受支持。")
        }
        val source = root.optJSONObject("source")
            ?: throw DailyInfoClientException("班车数据缺少来源。")
        val sourceName = requiredString(source, "name", 100)
        val sourcePage = trustedURL(requiredString(source, "page_url", 500))
            ?: throw DailyInfoClientException("班车来源地址不受信任。")
        val records = root.optJSONArray("items")
            ?: throw DailyInfoClientException("班车数据格式不正确。")
        if (records.length() > 100) throw DailyInfoClientException("班车通知数量异常。")
        val notices = buildList {
            for (index in 0 until records.length()) {
                val record = records.optJSONObject(index) ?: continue
                parseNotice(record)?.let(::add)
            }
        }.sortedWith(compareByDescending(ShuttleBusNotice::publishedAt).thenBy { it.id })
        return ShuttleBusSnapshot(
            generatedAt = requiredString(root, "generated_at", 80),
            status = requiredString(root, "status", 30).takeIf { it in setOf("healthy", "stale") }
                ?: throw DailyInfoClientException("班车服务状态无效。"),
            sourceName = sourceName,
            sourcePage = sourcePage,
            notices = notices,
        )
    }

    private fun parseNotice(record: JSONObject): ShuttleBusNotice? {
        val id = optionalString(record, "id", 160) ?: return null
        val title = optionalString(record, "title", 300) ?: return null
        val publishedAt = optionalString(record, "published_at", 10)
            ?.takeIf(StrictContractDate::isValid) ?: return null
        val schedules = record.optJSONArray("schedules")?.let { values ->
            buildList {
                for (index in 0 until minOf(values.length(), 24)) {
                    values.optJSONObject(index)?.let(::parseSchedule)?.let(::add)
                }
            }
        }.orEmpty()
        val stops = record.optJSONArray("stops")?.let { values ->
            buildList {
                for (index in 0 until minOf(values.length(), 12)) {
                    val stop = values.optJSONObject(index) ?: continue
                    val campus = optionalString(stop, "campus", 80) ?: continue
                    val location = optionalString(stop, "location", 160) ?: continue
                    add(campus to location)
                }
            }
        }.orEmpty()
        val notes = record.optJSONArray("notes")?.let { values ->
            buildList {
                for (index in 0 until minOf(values.length(), 8)) {
                    values.optString(index).trim()
                        .takeIf { it.isNotEmpty() && it.length <= 300 }
                        ?.let(::add)
                }
            }
        }.orEmpty()
        return ShuttleBusNotice(
            id = id,
            title = title,
            publishedAt = publishedAt,
            sourceURL = optionalString(record, "source_url", 500)?.let(::trustedURL),
            kind = optionalString(record, "kind", 40).orEmpty(),
            parseStatus = optionalString(record, "parse_status", 40).orEmpty(),
            stops = stops,
            notes = notes,
            schedules = schedules,
        )
    }

    private fun parseSchedule(record: JSONObject): ShuttleBusSchedule? {
        val periodRecord = record.optJSONObject("period") ?: JSONObject()
        val startDate = optionalString(periodRecord, "start_date", 10)
            ?.takeIf(StrictContractDate::isValid)
        val endDate = optionalString(periodRecord, "end_date", 10)
            ?.takeIf(StrictContractDate::isValid)
        if (startDate != null && endDate != null && endDate < startDate) return null
        val rows = record.optJSONArray("rows")?.let { values ->
            buildList {
                for (index in 0 until minOf(values.length(), 100)) {
                    values.optJSONObject(index)?.let(::parseRow)?.let(::add)
                }
            }
        }.orEmpty().sortedBy(ShuttleBusRow::departureTime)
        return ShuttleBusSchedule(
            period = ShuttleBusPeriod(
                label = optionalString(periodRecord, "label", 160) ?: "通知所示时段",
                startDate = startDate,
                endDate = endDate,
            ),
            from = optionalString(record, "from", 80),
            to = optionalString(record, "to", 80),
            parseStatus = optionalString(record, "parse_status", 40).orEmpty(),
            rows = rows,
        )
    }

    private fun parseRow(record: JSONObject): ShuttleBusRow? {
        val departureTime = optionalString(record, "departure_time", 5)
            ?.takeIf(timePattern::matches) ?: return null
        val servicesRecord = record.optJSONObject("services") ?: return null
        val services = weekdays.associateWith { weekday ->
            when (val value = servicesRecord.opt(weekday)) {
                null, JSONObject.NULL -> null
                is JSONObject -> {
                    val vehicle = optionalString(value, "vehicle", 40) ?: return@associateWith null
                    val count = value.optInt("count", 0)
                    if (count !in 1..20) null else ShuttleBusService(vehicle, count)
                }
                else -> null
            }
        }
        return ShuttleBusRow(departureTime, services)
    }

    private fun requiredString(source: JSONObject, key: String, maximum: Int): String =
        optionalString(source, key, maximum)
            ?: throw DailyInfoClientException("班车数据格式不正确。")

    private fun optionalString(source: JSONObject, key: String, maximum: Int): String? =
        (source.opt(key) as? String)?.trim()
            ?.takeIf { it.isNotEmpty() && it.length <= maximum }

    private fun trustedURL(value: String): String? = runCatching {
        val uri = URI.create(value)
        value.takeIf {
            uri.scheme.equals("https", ignoreCase = true) &&
                !uri.host.isNullOrBlank() && uri.userInfo == null
        }
    }.getOrNull()
}

internal object ShuttleBusLogic {
    private val shanghai = TimeZone.getTimeZone("Asia/Shanghai")

    fun today(snapshot: ShuttleBusSnapshot, now: Calendar): TodayShuttlePresentation {
        val localNow = Calendar.getInstance(shanghai).apply { timeInMillis = now.timeInMillis }
        val today = SimpleDateFormat("yyyy-MM-dd", Locale.ROOT).apply {
            timeZone = shanghai
        }.format(localNow.time)
        val weekday = weekdayKey(localNow.get(Calendar.DAY_OF_WEEK))
        val selection = selectCurrentNotice(snapshot.notices, today)
        val notice = selection?.first
        val routes = selection?.second.orEmpty().mapNotNull { schedule ->
            val from = schedule.from ?: return@mapNotNull null
            val to = schedule.to ?: return@mapNotNull null
            TodayShuttleRoute(
                from = from,
                to = to,
                periodLabel = schedule.period.label,
                departures = schedule.rows.mapNotNull { row ->
                    row.services[weekday]?.let { service ->
                        TodayShuttleDeparture(row.departureTime, service.vehicle, service.count)
                    }
                },
            )
        }
        val departureCount = routes.sumOf { it.departures.size }
        val vehicleCount = routes.sumOf { route -> route.departures.sumOf { it.count } }
        val status = when {
            notice == null -> "未找到当前生效的班车时刻表"
            departureCount == 0 -> "今日暂无已安排班车"
            else -> "今日安排 $departureCount 个发车时刻 · $vehicleCount 辆车"
        }
        val currentTime = "%02d:%02d".format(
            Locale.ROOT,
            localNow.get(Calendar.HOUR_OF_DAY),
            localNow.get(Calendar.MINUTE),
        )
        val next = routes.flatMap { route ->
            route.departures.map { departure -> route to departure }
        }.filter { (_, departure) -> departure.time > currentTime }
            .minByOrNull { (_, departure) -> departure.time }
            ?.let { (route, departure) ->
                "下一班 ${departure.time} · ${route.from} → ${route.to}"
            }
        return TodayShuttlePresentation(
            status = status,
            nextDeparture = next ?: if (departureCount > 0) "今日班车已结束" else null,
            noticeTitle = notice?.title,
            noticeURL = notice?.sourceURL,
            stops = notice?.stops.orEmpty(),
            notes = notice?.notes.orEmpty(),
            routes = routes,
            isStale = snapshot.status == "stale",
        )
    }

    private fun selectCurrentNotice(
        notices: List<ShuttleBusNotice>,
        today: String,
    ): Pair<ShuttleBusNotice, List<ShuttleBusSchedule>>? {
        notices.filter { it.publishedAt <= today }.forEach { notice ->
            val parsed = notice.schedules.filter {
                it.parseStatus == "parsed" && it.rows.isNotEmpty()
            }
            if (parsed.isEmpty()) return@forEach
            val explicit = parsed.filter { schedule ->
                val period = schedule.period
                period.startDate != null && period.startDate <= today &&
                    (period.endDate == null || period.endDate >= today)
            }
            if (explicit.isNotEmpty()) {
                val latestStart = explicit.maxOf { it.period.startDate.orEmpty() }
                return notice to explicit.filter { it.period.startDate.orEmpty() == latestStart }
            }
            // The newest usable notice is authoritative. Unknown, upcoming or
            // ended periods are never presented as the currently effective
            // timetable, and an older notice is not resurrected in a gap.
            return null
        }
        return null
    }

    private fun weekdayKey(dayOfWeek: Int): String = when (dayOfWeek) {
        Calendar.MONDAY -> "monday"
        Calendar.TUESDAY -> "tuesday"
        Calendar.WEDNESDAY -> "wednesday"
        Calendar.THURSDAY -> "thursday"
        Calendar.FRIDAY -> "friday"
        Calendar.SATURDAY -> "saturday"
        else -> "sunday"
    }
}

internal class ShuttleBusClient(
    private val fetchPublicJson: (URI, String, String, Int) -> String =
        FixedPublicJsonTransport::fetch,
) {
    fun fetch(): ShuttleBusSnapshot = try {
        val payload = fetchPublicJson(
            URI.create(CalendarDailyInfoSources.shuttleBus),
            "https",
            "where-to-study.cn",
            CalendarDailyInfoSources.shuttlePayloadLimit,
        )
        ShuttleBusResponseParser.parse(payload)
    } catch (error: DailyInfoClientException) {
        throw DailyInfoClientException("班车信息获取失败：${error.message}", error)
    } catch (error: Exception) {
        throw DailyInfoClientException("班车信息获取失败。", error)
    }
}

internal class ShuttleBusRepository(
    private val client: ShuttleBusClient = ShuttleBusClient(),
    private val usesSampleData: Boolean = DailyCourseNotificationRuntimeMode.isUiTesting,
) {
    private val worker = Executors.newSingleThreadExecutor()
    private val handler = Handler(Looper.getMainLooper())
    private val observers = CopyOnWriteArrayList<() -> Unit>()
    private val loading = AtomicBoolean(false)
    private val closed = AtomicBoolean(false)
    @Volatile var snapshot: ShuttleBusSnapshot? = null
        private set
    @Volatile var error: String? = null
        private set
    @Volatile private var fetchedAtMillis: Long = Long.MIN_VALUE

    fun isLoading(): Boolean = loading.get()

    fun addObserver(observer: () -> Unit) {
        if (!closed.get()) observers.addIfAbsent(observer)
    }

    fun removeObserver(observer: () -> Unit) {
        observers.remove(observer)
    }

    fun load(force: Boolean = false) {
        val nowMillis = System.nanoTime() / 1_000_000L
        val hasFreshSnapshot = snapshot != null &&
            nowMillis - fetchedAtMillis in 0 until CACHE_MILLIS
        if (closed.get() || (!force && hasFreshSnapshot) ||
            !loading.compareAndSet(false, true)
        ) return
        error = null
        try {
            worker.execute {
                runCatching {
                    if (usesSampleData) sampleSnapshot() else client.fetch()
                }.onSuccess {
                    snapshot = it
                    fetchedAtMillis = System.nanoTime() / 1_000_000L
                }
                    .onFailure { error = it.message ?: "班车信息获取失败。" }
                loading.set(false)
                if (!closed.get()) handler.post {
                    if (!closed.get()) observers.toList().forEach { it() }
                }
            }
        } catch (_: RejectedExecutionException) {
            loading.set(false)
        }
    }

    fun close() {
        if (!closed.compareAndSet(false, true)) return
        observers.clear()
        handler.removeCallbacksAndMessages(null)
        worker.shutdownNow()
    }

    private fun sampleSnapshot(): ShuttleBusSnapshot {
        val now = Calendar.getInstance(TimeZone.getTimeZone("Asia/Shanghai"))
        val today = SimpleDateFormat("yyyy-MM-dd", Locale.ROOT).apply {
            timeZone = now.timeZone
        }.format(now.time)
        val services = mapOf(
            "monday" to ShuttleBusService("大巴", 1),
            "tuesday" to ShuttleBusService("大巴", 1),
            "wednesday" to ShuttleBusService("大巴", 1),
            "thursday" to ShuttleBusService("大巴", 1),
            "friday" to ShuttleBusService("大巴", 1),
            "saturday" to ShuttleBusService("大巴", 1),
            "sunday" to ShuttleBusService("大巴", 1),
        )
        return ShuttleBusSnapshot(
            generatedAt = "${today}T00:00:00+08:00",
            status = "healthy",
            sourceName = "北京邮电大学后勤部",
            sourcePage = "https://hq.bupt.edu.cn/tzgg.htm",
            notices = listOf(ShuttleBusNotice(
                id = "sample-shuttle",
                title = "示例两校区班车时刻表",
                publishedAt = today,
                sourceURL = "https://hq.bupt.edu.cn/tzgg.htm",
                kind = "regular_schedule",
                parseStatus = "parsed",
                stops = listOf("西土城路校区" to "教三楼西侧", "沙河校区" to "学生活动中心南侧"),
                notes = listOf("请提前五分钟候车。"),
                schedules = listOf(
                    ShuttleBusSchedule(
                        ShuttleBusPeriod("当前执行时段", today, null),
                        "西土城路校区", "沙河校区", "parsed",
                        listOf(ShuttleBusRow("08:30", services), ShuttleBusRow("13:30", services)),
                    ),
                    ShuttleBusSchedule(
                        ShuttleBusPeriod("当前执行时段", today, null),
                        "沙河校区", "西土城路校区", "parsed",
                        listOf(ShuttleBusRow("09:50", services), ShuttleBusRow("17:30", services)),
                    ),
                ),
            )),
        )
    }

    private companion object {
        const val CACHE_MILLIS = 5 * 60 * 1_000L
    }
}

File: native/apple/Sources/Shared/ShuttleBusClient.swift
import Foundation

enum ShuttleBusSources {
    static let api = URL(string: "https://where-to-study.cn/api/shuttle-bus")!
    static let sourcePage = URL(string: "https://hq.bupt.edu.cn/tzgg.htm")!
    static let maximumPayloadBytes = 4 * 1024 * 1024
}

enum ShuttleBusError: LocalizedError, Equatable, Sendable {
    case service(String)

    var errorDescription: String? {
        switch self {
        case let .service(message): message
        }
    }
}

struct ShuttleBusService: Equatable, Sendable {
    let vehicle: String
    let count: Int
}

struct ShuttleBusDeparture: Identifiable, Equatable, Sendable {
    let departureTime: String
    let service: ShuttleBusService

    var id: String { departureTime }
}

struct ShuttleBusPeriod: Hashable, Sendable {
    let label: String
    let startDate: String?
    let endDate: String?

    var id: String {
        [startDate ?? "unknown", endDate ?? "open", label].joined(separator: ":")
    }

    func contains(_ date: String) -> Bool {
        guard let startDate, startDate <= date else { return false }
        return endDate.map { date <= $0 } ?? true
    }
}

struct ShuttleBusSchedule: Identifiable, Equatable, Sendable {
    let period: ShuttleBusPeriod
    let from: String
    let to: String
    let parseStatus: String
    let rows: [ShuttleBusScheduleRow]

    var id: String { "\(period.id):\(from):\(to)" }
}

struct ShuttleBusScheduleRow: Equatable, Sendable {
    let departureTime: String
    let services: [String: ShuttleBusService]
}

struct ShuttleBusStop: Identifiable, Equatable, Sendable {
    let campus: String
    let location: String

    var id: String { campus }
}

struct ShuttleBusNotice: Identifiable, Equatable, Sendable {
    let id: String
    let title: String
    let publishedAt: String
    let sourceURL: URL?
    let kind: String
    let notes: [String]
    let stops: [ShuttleBusStop]
    let parseStatus: String
    let schedules: [ShuttleBusSchedule]
}

struct ShuttleBusSnapshot: Equatable, Sendable {
    let generatedAt: String
    let status: String
    let sourceName: String
    let sourcePage: URL
    let lastParsedNoticeID: String?
    let notices: [ShuttleBusNotice]
}

protocol ShuttleBusFetching: Sendable {
    func fetch() async throws -> ShuttleBusSnapshot
}

struct ShuttleBusClient: ShuttleBusFetching {
    typealias DataLoader = @Sendable () async throws -> Data

    private let dataLoader: DataLoader

    init() {
        dataLoader = Self.loadLiveData
    }

    init(dataLoader: @escaping DataLoader) {
        self.dataLoader = dataLoader
    }

    func fetch() async throws -> ShuttleBusSnapshot {
        try Self.parse(data: try await dataLoader())
    }

    static func parse(data: Data) throws -> ShuttleBusSnapshot {
        guard data.count <= ShuttleBusSources.maximumPayloadBytes else {
            throw ShuttleBusError.service("班车数据响应过大。")
        }
        let payload: ShuttleBusPayloadDTO
        do {
            payload = try JSONDecoder().decode(ShuttleBusPayloadDTO.self, from: data)
        } catch {
            throw ShuttleBusError.service("班车数据格式不正确。")
        }
        guard payload.schemaVersion == "1.0", payload.items.count <= 50 else {
            throw ShuttleBusError.service("班车数据接口版本或数量不受支持。")
        }
        guard parseISO8601(payload.generatedAt) != nil else {
            throw ShuttleBusError.service("班车数据更新时间不正确。")
        }

        let notices = payload.items.compactMap(validatedNotice).sorted {
            ($0.publishedAt, $0.id) > ($1.publishedAt, $1.id)
        }
        guard !notices.isEmpty else {
            throw ShuttleBusError.service("班车接口暂未返回有效通知。")
        }
        let pageURL = trustedHTTPSURL(payload.source.pageURL, host: "hq.bupt.edu.cn")
            ?? ShuttleBusSources.sourcePage
        return ShuttleBusSnapshot(
            generatedAt: payload.generatedAt,
            status: payload.status,
            sourceName: payload.source.name,
            sourcePage: pageURL,
            lastParsedNoticeID: payload.lastParsedNoticeID,
            notices: notices
        )
    }

    private static func validatedNotice(_ value: ShuttleBusNoticeDTO) -> ShuttleBusNotice? {
        guard !value.id.isEmpty,
              StrictContractDateParser.date(from: value.publishedAt) != nil,
              !value.title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
              value.schedules.count <= 16
        else { return nil }

        let schedules = value.schedules.compactMap(validatedSchedule)
        return ShuttleBusNotice(
            id: value.id,
            title: value.title,
            publishedAt: value.publishedAt,
            sourceURL: trustedHTTPSURL(value.sourceURL, host: "hq.bupt.edu.cn"),
            kind: value.kind,
            notes: Array(value.notes.filter { !$0.isEmpty }.prefix(12)),
            stops: Array(value.stops.compactMap { stop in
                guard !stop.campus.isEmpty, !stop.location.isEmpty else { return nil }
                return ShuttleBusStop(campus: stop.campus, location: stop.location)
            }.prefix(8)),
            parseStatus: value.parseStatus,
            schedules: schedules
        )
    }

    private static func validatedSchedule(_ value: ShuttleBusScheduleDTO) -> ShuttleBusSchedule? {
        guard value.parseStatus == "parsed",
              !value.from.isEmpty,
              !value.to.isEmpty,
              value.rows.count <= 64
        else { return nil }
        if let startDate = value.period.startDate,
           StrictContractDateParser.date(from: startDate) == nil { return nil }
        if let endDate = value.period.endDate,
           StrictContractDateParser.date(from: endDate) == nil { return nil }
        let rows = value.rows.compactMap { row -> ShuttleBusScheduleRow? in
            guard validTime(row.departureTime) else { return nil }
            var services = [String: ShuttleBusService]()
            for (key, wrappedService) in row.services {
                guard ShuttleBusTodayLogic.weekdayKeys.contains(key),
                      let service = wrappedService,
                      !service.vehicle.isEmpty,
                      (1 ... 20).contains(service.count)
                else { continue }
                services[key] = ShuttleBusService(vehicle: service.vehicle, count: service.count)
            }
            return ShuttleBusScheduleRow(
                departureTime: row.departureTime,
                services: services
            )
        }.sorted { $0.departureTime < $1.departureTime }
        guard !rows.isEmpty else { return nil }
        return ShuttleBusSchedule(
            period: ShuttleBusPeriod(
                label: value.period.label,
                startDate: value.period.startDate,
                endDate: value.period.endDate
            ),
            from: value.from,
            to: value.to,
            parseStatus: value.parseStatus,
            rows: rows
        )
    }

    private static func trustedHTTPSURL(_ value: String?, host: String) -> URL? {
        guard let value,
              let url = URL(string: value),
              url.scheme == "https",
              url.host == host,
              url.user == nil,
              url.password == nil
        else { return nil }
        return url
    }

    private static func validTime(_ value: String) -> Bool {
        let parts = value.split(separator: ":", omittingEmptySubsequences: false)
        guard parts.count == 2,
              parts[0].count == 2,
              parts[1].count == 2,
              let hour = Int(parts[0]),
              let minute = Int(parts[1])
        else { return false }
        return (0 ... 23).contains(hour) && (0 ... 59).contains(minute)
    }

    private static func parseISO8601(_ value: String) -> Date? {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: value) { return date }
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: value)
    }

    private static func loadLiveData() async throws -> Data {
        var request = URLRequest(url: ShuttleBusSources.api)
        request.timeoutInterval = 20
        request.cachePolicy = .reloadIgnoringLocalCacheData
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 20
        configuration.timeoutIntervalForResource = 25
        let session = URLSession(
            configuration: configuration,
            delegate: ShuttleBusRedirectDelegate(),
            delegateQueue: nil
        )
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse,
              http.statusCode == 200,
              http.url?.scheme == "https",
              http.url?.host == "where-to-study.cn"
        else {
            throw ShuttleBusError.service("班车接口暂时不可用。")
        }
        guard data.count <= ShuttleBusSources.maximumPayloadBytes else {
            throw ShuttleBusError.service("班车数据响应过大。")
        }
        return data
    }
}

final class ShuttleBusRedirectDelegate: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

enum ShuttleBusTodayLogic {
    static let weekdayKeys = Set([
        "monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday",
    ])

    static func weekdayKey(for date: Date, calendar: Calendar = .shanghai) -> String {
        switch calendar.component(.weekday, from: date) {
        case 1: "sunday"
        case 2: "monday"
        case 3: "tuesday"
        case 4: "wednesday"
        case 5: "thursday"
        case 6: "friday"
        default: "saturday"
        }
    }

    static func scheduleNotice(in snapshot: ShuttleBusSnapshot) -> ShuttleBusNotice? {
        if let latest = snapshot.notices.first,
           latest.schedules.contains(where: { !$0.rows.isEmpty }) {
            return latest
        }
        guard let id = snapshot.lastParsedNoticeID else { return nil }
        return snapshot.notices.first { $0.id == id }
    }

    static func activeSchedules(
        in snapshot: ShuttleBusSnapshot,
        on date: Date,
        calendar: Calendar = .shanghai
    ) -> [ShuttleBusSchedule] {
        let dateString = StrictContractDateParser.string(from: date, calendar: calendar)
        guard let notice = scheduleNotice(in: snapshot) else { return [] }
        return notice.schedules.filter { $0.period.contains(dateString) }
    }

    static func departures(
        for schedule: ShuttleBusSchedule,
        on date: Date,
        calendar: Calendar = .shanghai
    ) -> [ShuttleBusDeparture] {
        let weekday = weekdayKey(for: date, calendar: calendar)
        return schedule.rows.compactMap { row in
            guard let service = row.services[weekday] else { return nil }
            return ShuttleBusDeparture(departureTime: row.departureTime, service: service)
        }
    }
}

@MainActor
final class ShuttleBusStore: ObservableObject {
    @Published private(set) var snapshot: ShuttleBusSnapshot?
    @Published private(set) var isLoading = false
    @Published private(set) var errorMessage = ""

    private let client: any ShuttleBusFetching

    init(client: any ShuttleBusFetching = ShuttleBusClient()) {
        self.client = client
    }

    func load(force: Bool = false, sampleMode: Bool = false) async {
        guard !isLoading, force || snapshot == nil else { return }
        if sampleMode {
            snapshot = ShuttleBusSampleData.snapshot()
            errorMessage = ""
            return
        }
        isLoading = true
        errorMessage = ""
        defer { isLoading = false }
        do {
            snapshot = try await client.fetch()
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

enum ShuttleBusSampleData {
    static func snapshot(now: Date = .now, calendar: Calendar = .shanghai) -> ShuttleBusSnapshot {
        let today = StrictContractDateParser.string(from: now, calendar: calendar)
        let end = calendar.date(byAdding: .day, value: 30, to: now) ?? now
        let period = ShuttleBusPeriod(
            label: "示例运行时段",
            startDate: today,
            endDate: StrictContractDateParser.string(from: end, calendar: calendar)
        )
        let weekdayServices = Dictionary(uniqueKeysWithValues: ShuttleBusTodayLogic.weekdayKeys.map {
            ($0, ShuttleBusService(vehicle: "大巴", count: 1))
        })
        let rows = ["06:30", "08:30", "12:00", "17:30"].map {
            ShuttleBusScheduleRow(departureTime: $0, services: weekdayServices)
        }
        let notice = ShuttleBusNotice(
            id: "sample-shuttle",
            title: "示例：两校区班车运行安排",
            publishedAt: today,
            sourceURL: nil,
            kind: "regular_schedule",
            notes: ["示例数据不连接班车服务。"],
            stops: [
                ShuttleBusStop(campus: "西土城路校区", location: "教三楼西侧"),
                ShuttleBusStop(campus: "沙河校区", location: "学生活动中心南侧"),
            ],
            parseStatus: "parsed",
            schedules: [
                ShuttleBusSchedule(
                    period: period,
                    from: "西土城路校区",
                    to: "沙河校区",
                    parseStatus: "parsed",
                    rows: rows
                ),
                ShuttleBusSchedule(
                    period: period,
                    from: "沙河校区",
                    to: "西土城路校区",
                    parseStatus: "parsed",
                    rows: rows
                ),
            ]
        )
        return ShuttleBusSnapshot(
            generatedAt: "2026-01-01T00:00:00+08:00",
            status: "healthy",
            sourceName: "示例数据",
            sourcePage: ShuttleBusSources.sourcePage,
            lastParsedNoticeID: notice.id,
            notices: [notice]
        )
    }
}

private struct ShuttleBusPayloadDTO: Decodable {
    let schemaVersion: String
    let generatedAt: String
    let status: String
    let source: ShuttleBusSourceDTO
    let lastParsedNoticeID: String?
    let items: [ShuttleBusNoticeDTO]

    enum CodingKeys: String, CodingKey {
        case schemaVersion = "schema_version"
        case generatedAt = "generated_at"
        case status, source, items
        case lastParsedNoticeID = "last_parsed_notice_id"
    }
}

private struct ShuttleBusSourceDTO: Decodable {
    let name: String
    let pageURL: String?

    enum CodingKeys: String, CodingKey {
        case name
        case pageURL = "page_url"
    }
}

private struct ShuttleBusNoticeDTO: Decodable {
    let id: String
    let title: String
    let publishedAt: String
    let sourceURL: String?
    let kind: String
    let notes: [String]
    let stops: [ShuttleBusStopDTO]
    let parseStatus: String
    let schedules: [ShuttleBusScheduleDTO]

    enum CodingKeys: String, CodingKey {
        case id, title, kind, notes, stops, schedules
        case publishedAt = "published_at"
        case sourceURL = "source_url"
        case parseStatus = "parse_status"
    }
}

private struct ShuttleBusStopDTO: Decodable {
    let campus: String
    let location: String
}

private struct ShuttleBusScheduleDTO: Decodable {
    let period: ShuttleBusPeriodDTO
    let from: String
    let to: String
    let parseStatus: String
    let rows: [ShuttleBusScheduleRowDTO]

    enum CodingKeys: String, CodingKey {
        case period, from, to, rows
        case parseStatus = "parse_status"
    }
}

private struct ShuttleBusPeriodDTO: Decodable {
    let label: String
    let startDate: String?
    let endDate: String?

    enum CodingKeys: String, CodingKey {
        case label
        case startDate = "start_date"
        case endDate = "end_date"
    }
}

private struct ShuttleBusScheduleRowDTO: Decodable {
    let departureTime: String
    let services: [String: ShuttleBusServiceDTO?]

    enum CodingKeys: String, CodingKey {
        case departureTime = "departure_time"
        case services
    }
}

private struct ShuttleBusServiceDTO: Decodable {
    let vehicle: String
    let count: Int
}
